### PR TITLE
15291 hide sole proprietor change button

### PR DESCRIPTION
**Pull request by jonathan-longe**
*Issue #:* /bcgov/entity#15291

The business has requested that non-staff users should not be able to modify the sole proprietor where the sole proprietor is an organization.  This PR hides the appropriate change button and updates the related help text.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: tests/unit/PeopleAndRoles.spec.ts
Comment (severinbeauvais) [inline tests/unit/PeopleAndRoles.spec.ts:473]: You can be specific here: a non-individual must be an organization.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:1069]: Is this a better syntax then `getOrgPeople(state)`?

I think you don't need the extra space between the conditional object and the array index -- I think this works too: `getOrgPeople[0]?.roles[0]?.roleType`.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:1071]: Please simplify this negative-negative logic. It may be better to use a variable for the second part (isSoleProp && isOrganization && isProprietor) and then return "isRoleStaff && isNewVariable".

Another simplification: if this is a SP then the officer must necessarily be a proprietor. (That's the definition of Proprietor. Likewise, a Partner only exists for a GP.)

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:1063]: Add blank line above.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:1068]: Since this getter is only used in 1 place, can it moved to that component (or its mixin)?

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (severinbeauvais) [inline src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:286]: As on line 299 and 300, what do you think about labelling this "canEdit" ?

I'm thinking "separation of concerns" -- this template doesn't need to know about SP or Proprietor or that the edit label is "Change". KISS here, and in the script you can explain the reason for the logic.
This works well with moving the getter from the store to this component (or its mixin) -- the store doesn't need to ko

File: tests/unit/PeopleAndRoles.spec.ts
Comment (jonathan-longe) [reply tests/unit/PeopleAndRoles.spec.ts:473]: Good call, I'll change it.

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1069]: With regards to your first question, yes!  I think it's always better to write pure functions where all dependencies are injected.  Testing functions where the dependencies are not injected is more difficult and refactoring is more difficult.

With regards to your second point, good to know, I'll fix it.

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1071]: Good points, I'll simplify it as you propose.

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1063]: Will do

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1068]: My recommendation would be that all business logic belongs in the store.  Placing the logic in a component -- even if it's only used once -- means that we're more likely to create a duplicate when we need identical logic for another component.   I'm not crazy about the idea of extending mixins after you said that they weren't supported in Vue3.

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (jonathan-longe) [reply src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:286]: Let's hold off deciding on an appropriate name until we've decided on the appropriate location -- getter vs component.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [reply src/store/getters/state-getters.ts:1068]: I might be wrong on the Vue3 mixin thing. Maybe they just need to be handled differently.

I generally agree with your point, but the "should I show this edit button" logic is very specific to the ListPeopleAndRoles component, so I still think it should be moved there instead of in the store. Not a deal breaker though.

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (severinbeauvais) [reply src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:286]: Hehehe

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1069]: Let's discuss the business logic issue further.  It would be great to have a clear vision of the end goal.

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts]: I really like the simplification you proposed!

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [reply src/store/getters/state-getters.ts:1069]: I'd like to see a code architecture where it's easy to create / drop in a new component without having to know much about the architecture, and it's easy to create a new thing (eg, filing) by simply putting existing components on a new page (view).

I'd like to see a relatively "dumb" UI, with as little business logic as possible.

And I like the idea of "independent" components... Maybe call them "smart" components (vs dumb), where most of what they do is localized and their interface to the rest of the system is clear / well defined. (I know this seems to go against the "store does everything" model. But maybe there's nice middle ground.)

I like the "tier" idea and separation of concerns, which to me keeps things independent and well defined. (It also makes unit testing easier.)

What do you think?

Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1069]: I think the architecture that I wrote for Roadsafety is what you're describing.  Form components (made up of one or more questions) could dropped into any new form and the only wiring required was the store location and the display options.  They had local smarts.  For example, they could display a spinner when busy but the component logic was encapsulated. The parent didn't need to know how the inner workings behaved and new fields could be added the shared component without changing the parent. I'll tee-up a meeting for us to discuss the implementation details.

Comment (severinbeauvais) [reply src/store/getters/state-getters.ts:1069]: Jonathan and I discussed this larger architectural question in various 1:1 meetings (esp in the context of the new Consent for Continuation Out filing, and re: app startup in App.vue, and mixins vs store modules).

Anyway, I'm OK with the implementation above since it's a temporary change, and long-term we'll figure out our ideal model so let's not get side tracked by that right now.

Resolving this conversation.

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (severinbeauvais) [inline src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:286]: If NOT hide ?

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (severinbeauvais) [inline src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:379]: Not needed.

File: src/store/getters/state-getters.ts
Comment (severinbeauvais) [inline src/store/getters/state-getters.ts:1064]: Switch the logic :)

File: tests/unit/PeopleAndRoles.spec.ts
Comment (severinbeauvais) [inline tests/unit/PeopleAndRoles.spec.ts:480]: You don't need the id for the getter, so prob better to leave it out. Also below.

**Review by severinbeauvais — APPROVED** [review]
GTG when you've sorted the logic.

Remember to update the app version in both PRs.

Would be nice to have other review approvals :)

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (jonathan-longe) [reply src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:286]: Ah, good thinking!

File: src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue
Comment (jonathan-longe) [reply src/components/common/PeopleAndRoles/ListPeopleAndRoles.vue:379]: Fixed

File: src/store/getters/state-getters.ts
Comment (jonathan-longe) [reply src/store/getters/state-getters.ts:1064]: Will do

File: tests/unit/PeopleAndRoles.spec.ts
Comment (jonathan-longe) [reply tests/unit/PeopleAndRoles.spec.ts:480]: Nice, good catch

**Review by severinbeauvais — APPROVED** [review]

**Review by severinbeauvais — COMMENTED** [review]
Looks good but this PR is against main branch. If you intend to fix build 3.9.15 then you'll have to rebase.

**Comment by jonathan-longe** [comment]
> Looks good but this PR is against main branch. If you intend to fix build 3.9.15 then you'll have to rebase.

Pull #448 was created to hotfix build 3.9.15; this PR was created to ensure the hotfix was also incorporated into the main branch.

**Pull request merged**
